### PR TITLE
[aoti-et] Add an ASR runner and an Whisper example to showcase how to use it

### PR DESCRIPTION
**Pull request by larryliu0820**
**Key Changes:**

*   Create new ASR runner extension in `extension/asr/runner/` with reusable runner components (runner.h/cpp)
*   Update CMake configuration files to support ASR runner builds (executorch-config.cmake, default.cmake, llm.cmake)
*   Add new Whisper model example in `examples/models/whisper/` with CMake build, README, and main.cpp runner
*   Bump optimum-executorch commit pin for Whisper support
*   Update CUDA CI workflow for testing

This change enables automatic speech recognition (ASR) capabilities in ExecuTorch with Whisper as the first supported model, following a similar pattern to the existing LLM runner infrastructure.

### CODE REVIEW

File: extension/asr/runner/runner.cpp [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+}
+
+Error AsrRunner::load_tokenizer() {
+  if (tokenizer_ && tokenizer_->is_loaded()) {
+    return Error::Ok;
+  }
+
+  auto tokenizer =
+      ::executorch::extension::llm::load_tokenizer(tokenizer_path_);
+  if (!tokenizer) {
```
Comment (Gasoonjia) [inline extension/asr/runner/runner.cpp]: use `ET_CHECK_OR_RETURN_ERROR`?

**Review by Gasoonjia — COMMENTED** [review]
LGTM on my side. I will let @jackzhxng give the stamp about asr_runner.

**Review by Gasoonjia — APPROVED** [review]
overall LGTM

File: extension/asr/runner/runner.h [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+ * Runner that owns a ASR model encoder + decoder pair exported as a single
+ * ExecuTorch module. A good example is Whisper
+ * (https://huggingface.co/openai/whisper-small)
+ *
+ * The module is expected to expose two callable methods:
+ *  - "encoder": processes precomputed audio features into encoder states.
+ *  - "text_decoder": consumes the decoder input ids, encoder output and cache
+ *    positions to autoregressively generate logits.
+ */
+class AsrRunner {
```
Comment (mergennachin) [inline extension/asr/runner/runner.h]: experimental

File: examples/models/whisper/README.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+This directory hosts a lightweight C++ helper that drives Whisper models
+exported to ExecuTorch. The `AsrRunner` owns the `Module` instance that
+wraps a bundled `.pte` program and optional `.ptd` weight file, loads the
+`encoder` and `text_decoder` methods, and exposes a `transcribe()` loop that
+streams decoded text pieces through a callback.
+
+The runner assumes:
+- `model.pte` contains both Whisper encoder and decoder entry points named
+  `encoder` and `text_decoder`.
+- External parameters (for example KV cache blocks) are stored in a companion
```
Comment (mergennachin) [inline examples/models/whisper/README.md]: KV cache blocks are stored in ptd??

File: examples/models/whisper/README.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+  `model.ptd`.
+- A tokenizer JSON compatible with the ExecuTorch tokenizers shim is available.
+
+Audio preprocessing is not part of the runner itself. To transform raw audio
+into the mel features expected by the encoder, reuse the pattern in
+`examples/models/voxtral/multimodal.cpp`, which loads a `preprocessor.pte`
+module to generate the spectrogram tensor.
+
+## Build
+
```
Comment (mergennachin) [inline examples/models/whisper/README.md:22]: Mention that this is CUDA build 

I presume we can/should also add XNNPACK as well as Metal versions

File: examples/models/whisper/README.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+if (!tokens_result.ok()) {
+  ET_LOG(Error, "Transcription failed: %d", static_cast<int>(tokens_result.error()));
+}
+```
+
+`transcribe()` returns the full token history (prompt + generated tokens) and
+invokes the callback every time a new token is emitted. Provide a non-empty
+`decoder_input_ids` vector if you want to seed the decoder with a custom prompt,
+and override `AsrTranscribeConfig::eos_token_ids` when the model exposes
+custom termination ids.
```
Comment (mergennachin) [inline examples/models/whisper/README.md]: Move this to extensions/asr/runner documentation instead.

File: .ci/scripts/export_model_cuda_artifact.sh [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+case "$HF_MODEL" in
+  mistralai/Voxtral-Mini-3B-2507)
+    MODEL_NAME="voxtral"
+    TASK="multimodal-text-to-text"
+    MAX_SEQ_LEN="1024"
+    EXTRA_PIP="mistral-common librosa"
+    PREPROCESSOR_FEATURE_SIZE="128"
+    PREPROCESSOR_OUTPUT="voxtral_preprocessor.pte"
+    ;;
+  openai/whisper-small)
```
Comment (mergennachin) [inline .ci/scripts/export_model_cuda_artifact.sh:65]: does large model work?

File: extension/asr/runner/runner.h [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+#include <executorch/extension/llm/runner/llm_runner_helper.h>
+#include <executorch/extension/llm/runner/stats.h>
+#include <executorch/extension/llm/sampler/sampler.h>
+#include <executorch/extension/module/module.h>
+#include <executorch/extension/tensor/tensor_ptr.h>
+#include <executorch/runtime/core/error.h>
+#include <executorch/runtime/core/result.h>
+#include <pytorch/tokenizers/tokenizer.h>
+
+namespace executorch::extension::asr {
```
Comment (mergennachin) [inline extension/asr/runner/runner.h:26]: Once this lands, we can renew this PR and add language bindings

https://github.com/pytorch/executorch/pull/13525/files#diff-22a92f60f6a05a9cf1de00b0174c7d548066f60c935cd47dc3e288463b046149

File: .github/workflows/cuda.yml [hunk 8]
```diff
@@ -330,182 +205,49 @@ jobs:
...
+              - "quantized-int4-weight-only"
     with:
       timeout: 90
       runner: linux.g5.4xlarge.nvidia.gpu
       gpu-arch-type: cuda
       gpu-arch-version: 12.6
       use-custom-docker-registry: false
       submodules: recursive
-      download-artifact: ${{ matrix.format.artifact }}
+      download-artifact: ${{ matrix.hf_model.repo }}-${{ matrix.hf_model.name }}-cuda-${{ matrix.quant }}
```
Comment (jackzhxng) [inline .github/workflows/cuda.yml]: Nit: stick to `matrix.model`

File: extension/asr/runner/runner.h [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+  }
+
+  /**
+   * Sample the next token from the logits tensor.
+   * @param logits_tensor The logits tensor.
+   * @param temperature The temperature parameter used to control randomness in
+   * sampling.
+   * @return The next token.
+   */
+  inline int32_t logits_to_token(
```
Comment (mergennachin) [inline extension/asr/runner/runner.h]: Does this need to be a member function? Can it live in as a standalone util method?

Comment (mergennachin) [reply extension/asr/runner/runner.h]: Or part of this Whisper's main.cpp instead? 

I want to keep the API surface in extension to be as little as possible for starters

**Review by jackzhxng — APPROVED** [review]

File: examples/models/whisper/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+if(EXECUTORCH_BUILD_CUDA)
+  find_package(CUDAToolkit REQUIRED)
+  list(APPEND _link_libraries aoti_cuda)
+  executorch_target_link_options_shared_lib(aoti_cuda)
+endif()
+
+if(EXECUTORCH_BUILD_METAL)
+  list(APPEND _link_libraries metal_backend)
+  executorch_target_link_options_shared_lib(metal_backend)
+endif()
```
Comment (mergennachin) [inline examples/models/whisper/CMakeLists.txt:79]: Is metal supported? 

@manuelcandales

**Review by mergennachin — CHANGES_REQUESTED** [review]
Great work @larryliu0820 

A few comments before landing

File: .github/workflows/cuda.yml [hunk 11]
```diff
@@ -330,182 +205,49 @@ jobs:
...
-          echo "Expected output 'chip' not found in output"
-          exit 1
-        fi
-
-        if [ $EXIT_CODE -ne 0 ]; then
-          echo "Unexpected exit code: $EXIT_CODE"
-          exit $EXIT_CODE
-        fi
-        echo "::endgroup::"
+        source .ci/scripts/test_model_cuda_e2e.sh "${{ matrix.hf_model.repo }}" "${{ matrix.hf_model.name }}" "${{ matrix.quant }}" "${RUNNER_ARTIFACT_DIR}"
```
Comment (jackzhxng) [inline .github/workflows/cuda.yml]:
```suggestion
        source .ci/scripts/test_model_cuda_e2e.sh "${{ matrix.hf_model.repo }}/${{ matrix.hf_model.name }}" "${{ matrix.quant }}" "${RUNNER_ARTIFACT_DIR}"
```

File: .ci/scripts/test_model_cuda_e2e.sh [hunk 12]
```diff
@@ -0,0 +1,205 @@
+#!/bin/bash
```
Comment (mergennachin) [inline .ci/scripts/test_model_cuda_e2e.sh:1]: great work to refactor the ci scripts

File: extension/asr/runner/runner.h [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+ * The module is expected to expose two callable methods:
+ *  - "encoder": processes precomputed audio features into encoder states.
+ *  - "text_decoder": consumes the decoder input ids, encoder output and cache
+ *    positions to autoregressively generate logits.
+ */
+class AsrRunner {
+ public:
+  AsrRunner(
+      std::string module_path,
+      std::string data_path,
```
Comment (mergennachin) [inline extension/asr/runner/runner.h]: data_path can be optional no? For other backends, like Qualcomm

File: extension/asr/runner/runner.h [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+  /**
+   * Loads the module, validates required methods and initialises tokenizer.
+   */
+  ::executorch::runtime::Error load();
+
+  /**
+   * Executes an end-to-end transcription cycle.
+   *
+   * @param preprocessed_features Audio features already processed by a
+   * preprocessor module (see voxtral example).
```
Comment (mergennachin) [inline extension/asr/runner/runner.h]: Since this a first-class API now, instead of saying "see voxtral example", actually specify what's expected from this input. For example, is it [batch, time, features] or [batch, features, time]?

File: extension/asr/runner/runner.h [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+  /**
+   * Returns true when the module and tokenizer are ready for inference.
+   */
+  bool is_loaded() const;
+
+  /**
+   * Loads the module, validates required methods and initialises tokenizer.
+   */
+  ::executorch::runtime::Error load();
+
```
Comment (mergennachin) [inline extension/asr/runner/runner.h:76]: Don't you need an explicit reset() method similar to LLmrunner?

File: .ci/scripts/test_model_cuda_e2e.sh [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+    TOKENIZER_FILE="tekken.json"
+    AUDIO_URL="https://github.com/voxserv/audio_quality_testing_samples/raw/refs/heads/master/testaudio/16000/test01_20s.wav"
+    AUDIO_FILE="poem.wav"
+    IMAGE_PATH=""
+    ;;
+  openai/whisper-small)
+    MODEL_NAME="whisper"
+    RUNNER_TARGET="whisper_runner"
+    RUNNER_PATH="whisper"
+    EXPECTED_OUTPUT="Mr. Quilter"
```
Comment (mergennachin) [inline .ci/scripts/test_model_cuda_e2e.sh]: Make it a full transcription "Mr. Quilter is the apostle of the middle classes"

File: examples/models/whisper/README.md (hunk 3 above)
Comment (larryliu0820) [reply examples/models/whisper/README.md]: Haha I need to change these

File: .ci/scripts/export_model_cuda_artifact.sh (hunk 6 above)
Comment (larryliu0820) [reply .ci/scripts/export_model_cuda_artifact.sh:65]: No not yet

File: extension/asr/runner/runner.h (hunk 15 above)
Comment (larryliu0820) [reply extension/asr/runner/runner.h:76]: No I don't manage kv cache position internally and whisper doesn't have the "multiturn" use case

File: extension/asr/runner/runner.h (hunk 9 above)
Comment (larryliu0820) [reply extension/asr/runner/runner.h]: Yeah I can extract out into a util

**Review by mergennachin — APPROVED** [review]
Thanks
